ignite/pkg/cache/cache.go

```diff
@@ -0,0 +1,151 @@
+	"github.com/ignite-hq/cli/ignite/pkg/xfilepath"
+	bolt "go.etcd.io/bbolt"
+)
+
+const cacheFolderName = "cache"
+const dbName = "ignite_cache.db"
+
+var (
+	// This is the top level folder where each chain's cache will be stored
+	// When using NewChainStorage, each chain gets a separate folder
+	cacheFolder = xfilepath.Join(chainconfig.ConfigDirPath, xfilepath.Path(cacheFolderName))
+
+	// We need to make sure we don't open multiple connections to the same cache file, so we keep track of them here
+	dbInstances = make(map[string]*bolt.DB)
+)
+
+// Storage is meant to be passed around and used by the New function (which provides namespacing and type-safety)
+type Storage struct {
+	db *bolt.DB
+}
+
+// Cache is a namespaced and type-safe key-value store
+type Cache[T any] struct {
+	storage   Storage
+	namespace string
+}
+
+// NewChainStorage creates a separate cache storage for a chain with chainName
+// It is safe to call multiple times
+func NewChainStorage(chainName string) (Storage, error) {
+	dir, err := xfilepath.Join(cacheFolder, xfilepath.Path(chainName))()
+	if err != nil {
+		return Storage{}, err
+	}
+
+	return NewStorage(dir)
+}
+
+// NewStorage create a local db file (if it doesn't exist) and opens a connection to it (or reuses an existing one)
+// Usually NewChainStorage is more appropriate to use since it makes it possible to clear the cache for a particular project
+// It is safe to call multiple times
+func NewStorage(dir string) (Storage, error) {
+	db, ok := dbInstances[dir]
+	if ok {
+		return Storage{db}, nil
+	}
+
+	if err := os.MkdirAll(dir, 0755); err != nil {
+		return Storage{}, err
+	}
+
+	storagePath := filepath.Join(dir, dbName)
+	db, err := bolt.Open(storagePath, 0640, nil)
+	if err != nil {
+		return Storage{}, err
+	}
+
+	dbInstances[dir] = db
+	return Storage{db}, nil
+}
+
+// New creates a namespaced and typesafe key-value Cache
+func New[T any](storage Storage, namespace string) Cache[T] {
+	return Cache[T]{
+		storage:   storage,
+		namespace: namespace,
+	}
+}
+
+// Clear deletes all namespaces and cached values
+func (s Storage) Clear() error {
+	return s.db.Update(func(tx *bolt.Tx) error {
+		return tx.ForEach(func(name []byte, _ *bolt.Bucket) error {
+			return tx.DeleteBucket(name)
+		})
+	})
+}
+
+// Put sets key to value within the namespace
+// If the key already exists, it will be overwritten
+func (c Cache[T]) Put(key string, value T) error {
+	var buf bytes.Buffer
+	encoder := gob.NewEncoder(&buf)
+	if err := encoder.Encode(value); err != nil {
+		return err
+	}
+	result := buf.Bytes()
+
+	return c.storage.db.Update(func(tx *bolt.Tx) error {
+		b, err := tx.CreateBucketIfNotExists([]byte(c.namespace))
+		if err != nil {
+			return err
+		}
+		return b.Put([]byte(key), result)
+	})
+}
+
+// Get fetches the value of key within the namespace.
+// If no value exists, it will return found == false
+func (c Cache[T]) Get(key string) (val T, found bool, err error) {
```

Comment:
As you can see from the signature this returns 3 values, which I was hoping to avoid. I tried many different designs for the cache store, but I found this to be the cleanest in the end. I am very open to suggestions here.

Another thing I considered was if we want to actually swallow errors here or not (and just return found=false instead). It would simplify the UX of the cache module significantly, but potentially hide errors too.

Reply:
Yeah, I did consider that as well, but I personally found the resulting code not much cleaner from this, but not a lot worse either I suppose. Let me know if you want to change this or not.

Reply:
I like that. Making that change now!